### PR TITLE
screencast: fix monitor hotplug

### PR DESCRIPTION
**Pull request by columbarius**
This MR aims to fix the corrupted output list after a monitor hotplug.

Related to #76

### CODE REVIEW

File: src/screencast/wlr_screencast.c [hunk 1]
```diff
@@ -543,7 +543,11 @@ struct xdpw_wlr_output *xdpw_wlr_output_find(struct xdpw_screencast_context *ctx
 }
 
 static void wlr_remove_output(struct xdpw_wlr_output *out) {
+	free(out->name);
+	free(out->make);
+	free(out->model);
 	wl_list_remove(&out->link);
+	free(out);
 }
```
Comment (columbarius) [inline src/screencast/wlr_screencast.c:551]: Should out->xdg_output also be freed here? It's returned from [zxdg_output_manager_v1_get_xdg_output](https://github.com/emersion/xdg-desktop-portal-wlr/blob/master/protocols/xdg-output-unstable-v1.xml#L71).

**Comment by columbarius** [comment]
The second issue it the missing monitor name. This is the result of calling the [wlr_xdg_output_listener](https://github.com/emersion/xdg-desktop-portal-wlr/blob/master/src/screencast/wlr_screencast.c#L299) only on [xdpw_wlr_screencopy_init](https://github.com/emersion/xdg-desktop-portal-wlr/blob/master/src/screencast/wlr_screencast.c#L618) and not for any output added afterwards. Maybe we should add the call to the wlr_xdg_output_listener in [wlr_output_handle_done](https://github.com/emersion/xdg-desktop-portal-wlr/blob/master/src/screencast/wlr_screencast.c#L264) but we need to screencast_context for that and I don't know, if a roundtrip is needed.

**Comment by emersion** [comment]
In `wlr_registry_handle_remove`, there's no guarantee that the removed global is an output. We should check the return value of `xdpw_wlr_output_find`.

File: src/screencast/wlr_screencast.c (hunk 1 above)
Comment (emersion) [reply src/screencast/wlr_screencast.c:551]: Yes, via `zxdg_output_v1_destroy`.

Comment (emersion) [reply src/screencast/wlr_screencast.c:551]: We should also free the `wl_output`, via `wl_output_destroy`.

**Comment by emersion** [comment]
> Maybe we should add the call to the wlr_xdg_output_listener in wlr_output_handle_done but we need to screencast_context for that and I don't know, if a roundtrip is needed.

We don't need to wait for `wl_output.done`. Also, `wl_output.done` is sent when the output properties change, we don't want to re-create the `xdg_output` object when that happens.

An issue is that the order of globals is undefined, there's no guarantee that the `xdg_output_manager` will come before the `wl_output` globals.

I think we can create the `xdg_output` right after binding to the `wl_output`, assuming the `xdg_output_manager` is available. Then on init only create the `xdg_output` for each `wl_output` if it doesn't already exist.

File: src/screencast/wlr_screencast.c (hunk 1 above)
Comment (columbarius) [reply src/screencast/wlr_screencast.c]: thanks. added it.

**Comment by columbarius** [comment]
> > Maybe we should add the call to the wlr_xdg_output_listener in wlr_output_handle_done but we need to screencast_context for that and I don't know, if a roundtrip is needed.
> 
> We don't need to wait for `wl_output.done`. Also, `wl_output.done` is sent when the output properties change, we don't want to re-create the `xdg_output` object when that happens.
> 
> An issue is that the order of globals is undefined, there's no guarantee that the `xdg_output_manager` will come before the `wl_output` globals.
> 
> I think we can create the `xdg_output` right after binding to the `wl_output`, after checking that the `xdg_output_manager` is available. Then on init only create the `xdg_output` for each `wl_output` if it doesn't already exist.

Thanks for the explanation.

Current state: Output list doesn't corrupt anymore. Removing the output while screencasting will segfault xdpw. We could try to stop the corresponding screencast instances, when we remove the output, but that's probably  something for a followup mr.

**Review by emersion — APPROVED** [review]
LGTM, thanks!

**Pull request merged**